synapse/storage/state.py

```diff
@@ -585,7 +649,7 @@ def _get_state_for_groups(self, groups, types=None):
                 types_to_fetch = types
 
             group_to_state_dict = yield self._get_state_groups_from_groups(
-                missing_groups, types_to_fetch,
+                missing_groups, types_to_fetch, filtered_types
```

Comment:
At the risk of being stabbed, now that you've done all the work to update `_get_state_groups_from_groups`...

I don't think we're ever going to get here with `filtered_types` set without `types_to_fetch` being None (and hence filtered_types being ignored).

Unless you want to put in the work to make the DictionaryCache more intelligent, I suspect you need to pull the `filtered_types` support back out of `_get_state_groups_from_groups`.

Reply:
also: the `update the result` code below needs to take `filtered_types` into account.

Reply:
so to check i understand: you're saying that `_get_state_for_groups` deliberately loads all state from the DB if we are querying state_key wildcards in order to fully populate the cache.  So none of the filtering gets applied to the DB, so the `filtered_types` stuff is irrelevant and will never apply.

I'm not convinced - surely if you call `_get_state_for_groups` with types of `[('m.room.member', '@matthew:matrix.org'), ('m.room.member', '@richvdh:sw1v.org')]`, there are no wildcards present, and so the request will go all the way through to the DB - and you'd need `filtered_types` to be set to `['m.room.member']` to get the rest of the state.

Or are you saying that because `filtered_types` acts as wildcard (i.e. "please return me all state outside this list"), we should go and blindly load all state out of the DB and then apply the filtering in the `update the result` phase?

Reply:
i.e. something like https://gist.github.com/ara4n/55475cfe72c172e7df89b52eb20d25e3 ?

Reply:
actually, i'll just push it.

Reply:
> Or are you saying that because filtered_types acts as wildcard (i.e. "please return me all state outside this list"), we should go and blindly load all state out of the DB and then apply the filtering in the update the result phase?

yes, this.